apache/superset
# fix: added parameter to optionally rename aliases of aggregated columns used in GROUP BY statments

### SUMMARY

Fixed issue #28443 
Renamed `db_engine_specs` parameter `allows_alias_to_source_column` to `order_by_allows_alias_to_source_column` and added the new parameter `group_by_allows_alias_to_source_column`.

The previous parameter is used to tell the SQLA generator to rename aliases used in `ORDER BY` statements with aggregations, to ensure that the source column is referenced. Some engines (e.g. Drill) needs to be able to do the same thing for aliases in `GROUP BY` statements.

The new parameter is used to tell the SQLA generator to rename any alias of a source column that is used in an aggregation in a `GROUP BY` statement, to ensure that the source column is referenced.

Added documentation of the new `group_by_allows_alias_to_source_column` parameter, and fixed errors in the documentation of `order_by_allows_alias_to_source_column`/(previously)`allows_alias_to_source_column`

For example this query
```
SELECT length(n_name) AS n_name
...
GROUP BY length(n_name)
```
becomes
```
SELECT length(n_name) AS n_name__
...
GROUP BY length(n_name)
```

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
![Screenshot from 2024-05-12 19-37-24](https://github.com/apache/superset/assets/14278584/d83ba91c-a23a-4621-80ac-ac1a9320aca9)
![Screenshot from 2024-05-12 19-39-17](https://github.com/apache/superset/assets/14278584/6c02118f-b94e-43e3-9ccf-04de15517d8c)


### TESTING INSTRUCTIONS

1. Start Superset
2. Navigate to the Superset web application and login
3. Connect to a Drill database
4. Create a dataset of the Drill database
5. Create a chart from that dataset
6. Select visualization type Table with query mode Aggregate
7. Add two columns in the dimensions
8. Aggregate the data of one of the columns (e.g. length(column_a))
9. Press "view query"
10. The query should now contain a GROUP BY statement of an aggregation, and the alias of that aggregation should have "__" at the end of the name.
Example:
```
SELECT length(n_name) AS n_name__
...
GROUP BY length(n_name)
```

### ADDITIONAL INFORMATION


- [X] Has associated issue: Fixes #28443
